Add support for QOA (Quite OK Audio) format

DeeJayLSP opened the pull request:
Implements support for the Quite OK Audio format (QOA). Closes godotengine/godot-proposals#9133. More details on that proposal.

QOA was developed to (not exactly just this) be a better alternative to IMA-ADPCM for use in games according to [the article announcing it](https://phoboslab.org/log/2023/02/qoa-time-domain-audio-compression) (which unironically uses Godot as an example).

Briefly, this adds a way for developers to compress sound effects without relying on expensive decoding (MP3, Ogg Vorbis) or destroyed quality (IMA-ADPCM).

A deprecation notice could be added later for when someone tries to import WAV as IMA-ADPCM (getting godotengine/godot-proposals#4264, where QOA was first mentioned, closed too). Adding a converter from WAV was suggested there.

The final release template binaries (at least on Linux) have a size increase of only 24KiB, but it gets compensated if at least 0.18-0.35s of WAV sound gets converted to QOA.

`qoa.h` had to be patched to suppress a few warnings.

---
## Approaches

#### 1. QOA files can be imported directly
- Place a QOA file into the project folder and it will get imported directly.

#### 2. WAV files can be imported to use QOA internally
- Go to the FileSystem dock, click on a WAV file, open the Import tab and switch the importer from Microsoft WAV to Quite OK Audio. The file will be converted and imported as QOA.
- A file that was previously imported as Microsoft WAV will retain all of its custom properties when changing to Quite OK Audio.
___

## Caveats
- Lots of lines from ResourceImporterWAV are duplicated into ResourceImporterQOA. Something required to make the converter work. A general WAV converter could be made into a header and included on both WAV and QOA importers later without breaking compatibility;
- It uses AudioStreamPlaybackResampled, like MP3 and Vorbis. It's said that AudioStreamPlaybackWAV uses linear interpolation for resampling because it's cheaper to calculate, therefore being better for resampling lots of sound effects, while Resampled uses cubic, slightly more expensive, but some comments state that cubic should be fast enough nowadays. There are two open PRs related to resampling: one to make AudioStreamWAV inherit from Resampled and another that optimizes Resampled's cubic hermite resampling. Merging the former automatically dismisses this caveat;

## Issues to address in case this is accepted:
- ~~WAVs get imported as Quite OK Audio instead of Microsoft WAV by default;~~
    - Fixed by overriding `get_priority` to return 0.9 instead of the default value of 1.0;
- Reviewer demands;
    - Possible flaws in the code? Note that I made sure to fix every minor issue with it;
    - Does the way the patch is done satisfy reviewers?

Mickeon commented on modules/qoa/doc_classes/AudioStreamQOA.xml:
```diff
@@ -0,0 +1,47 @@
+		QOA audio stream driver.
+	</brief_description>
+	<description>
+		QOA audio stream driver. See [member data] if you want to load a QOA file at run-time.
```
Not convinced about the current wording but for future reference it may be worth linking to the website:
```suggestion
		[url=https://qoaformat.org/]QOA[/url] audio stream driver. See [member data] if you want to load a QOA file at run-time.
```

DeeJayLSP commented:
I know images in Godot have their importer somehow enforced to be Texture2D at first. Maybe the first issue could be fixed somehow like this?

Mickeon commented:
I believe it has something to do with [`get_import_order`](https://docs.godotengine.org/en/stable/classes/class_editorimportplugin.html#class-editorimportplugin-private-method-get-import-order). Try overriding that.

DeeJayLSP commented:
[get_import_order](https://docs.godotengine.org/en/stable/classes/class_editorimportplugin.html#class-editorimportplugin-private-method-get-import-order) is used to determine the order of certain importers.

For example, if an importer for a `.xy` file has an import order of 0, and one for a `.zw` file has an order of 1, the editor will import all `.xy` files before `.zw`.

The actual override would be for [get_priority](https://docs.godotengine.org/en/stable/classes/class_editorimportplugin.html#class-editorimportplugin-private-method-get-priority), which actually does what we want. I have overriden that to return 0.9 so ResourceImporterWAV gets a higher priority.

fire commented:
Is it possible for the priorities to be adjusted so they're integers (in a float)? Thoughts?

DeeJayLSP commented on modules/qoa/doc_classes/AudioStreamQOA.xml:
```diff
@@ -0,0 +1,47 @@
+		</member>
+		<member name="data" type="PackedByteArray" setter="set_data" getter="get_data" default="PackedByteArray()">
+			Contains the audio data in bytes.
+			You can load a file without having to import it beforehand using the code snippet below. Keep in mind that this snippet loads the whole file into memory and may not be ideal for huge files (hundreds of megabytes or more).
```
This is mostly copypasted from AudioStreamMP3.xml. If we should modify it, the changes should go there first.

Mickeon replied:
I have no suggestions documentation-wise, then. It's all copy-pasted from the others. It... definitely needs to be cleaned up at some point.

DeeJayLSP replied:
Well, not everything is copypasted. I would like to know what are the oddities.

DeeJayLSP commented:
> Is it possible for the priorities to be adjusted so they're integers (in a float)? Thoughts?
Seems like `get_priority()` can't return 0 or lower. For this to work properly we would have to make ResourceImporterWAV's return 2.

DeeJayLSP commented:
Should I remove BPM/Beat Count/Bar Beats? I don't think they are necessary since the format is intended for sound effects and Vorbis does a much better job compressing?

fire commented:
Wave cannot be deprecated since it is the majority of audio inputs. 

I don’t recommend removing the music code from QOA.

DeeJayLSP commented:
> Wave cannot be deprecated since it is the majority of audio inputs from libraries

WAV itself isn't getting deprecated. I'm just suggesting we deprecate IMA-ADPCM (which Godot itself cannot import) and it wouldn't apply before version 5, which isn't coming anytime soon.
> I don’t recommend removing the music code from QOA.

My excuse for doing so is because I don't think QOA is suitable for music at all. You can get a better quality even with MP3 at a lower bitrate and you don't need to play over 10 pieces at the same time.

fire commented:
If QOA is not suitable for music at all, then why is it suitable for sound effects?

Mickeon commented:
Never been a fan of the fact that basically every AudioStream has to "reimplement" and re-expose these music-related methods. Feels like blatant code duplication that could be avoided with restructuring, so I am not against removing them all from QOA until it is actually requested.

DeeJayLSP commented:
I've been making some changes to make importer and stream closer to WAV than music formats (MP3 and Vorbis) so it can have some feature parity.

The only thing left to do is looping, which still follows music formats and has been kinda hard to implement.

DeeJayLSP commented:
Apart from loops, everything seem to be in a good state.

I was unable to implement a functional `loop_end` property that didn't mess up the audio preview.

DeeJayLSP commented:
Added loop modes. Pingpong/backward loops seem quite hard to implement, so at least I left it in a way they could be implemented without breaking compatibility.

DeeJayLSP commented:
There seems to be an issue causing the returned sample to be `p_frames` samples ahead of `frames_mixed`, both on start and end. That's why I added the offset.

Time to find a fix for this.

DeeJayLSP commented:
Fixed the `p_frames` offset issue by making mix internal request only one frame to the decoder instead of `p_frames` frames.

This implementation is now in the best possible state I could find.

Mickeon commented:
I may suggest poking the Audio team on RocketChat. Right now, The sentiment for this may be that, because it can be done as an addon, it should not be in core, at least not until there's _a lot_ of demand.

AThousandShips commented on modules/qoa/resource_importer_qoa.cpp:
```diff
@@ -0,0 +1,489 @@
+#include "core/io/file_access.h"
+#include "core/io/marshalls.h"
+#include "core/io/resource_saver.h"
+#include "scene/resources/texture.h"
```
What are you using textures for in an audio importer?

DeeJayLSP replied:
It's some leftover from my previous implementation.

AThousandShips commented on modules/qoa/audio_stream_qoa.cpp:
```diff
@@ -0,0 +1,283 @@
+	qoa_desc qoad;
+	qoad.channels = 0;
+	qoad.samplerate = 0;
+	uint32_t ffp = qoa_decode_header(p_data.ptr(), QOA_MIN_FILESIZE, &qoad);
```
```suggestion
	uint32_t ffp = qoa_decode_header(p_data.ptr(), size, &qoad);
```
Or it will crash trying to read empty buffers, you can see this in the test runners

Calinou commented on doc/classes/ResourceImporterWAV.xml:
```diff
@@ -5,6 +5,7 @@
 	</brief_description>
 	<description>
 		WAV is an uncompressed format, which can provide higher quality compared to Ogg Vorbis and MP3. It also has the lowest CPU cost to decode. This means high numbers of WAV sounds can be played at the same time, even on low-end deviceS.
```
```suggestion
		WAV is an uncompressed format, which can provide higher quality compared to Ogg Vorbis and MP3. It also has the lowest CPU cost to decode. This means high numbers of WAV sounds can be played at the same time, even on low-end devices.
```

Calinou reviewed (commented):
Tested locally, it works as expected.

In terms of quality, I can't spot the difference with the uncompressed WAV source on my Audeze Maxwell headphones. There's no doubt this sounds much better than IDA-ADPCM.

With a 38-second 44,100 Hz, 16-bit stereo audio file, the uncompressed WAV is 6.5 MB and the corresponding imported QOA file (in `.godot/imported`) is 1.3 MB. This is a reduction of over 5 MB, which is much greater than the file size added to export templates by this PR. Even if you only had a few seconds of mono audio to compress to QOA, you'd still easily save the ~25 KB added by this PR.

My only concern is with the different import mode required, as we don't use this approach for any other audio format (unless the source format is different, like Ogg Vorbis and MP3). I don't have a good answer to this though, as refactoring it sounds difficult.

There's also the alternative of decoding Ogg Vorbis/MP3 to uncompressed audio at load-time which could work well, at the cost of requiring more memory (it would have to be something you enable on import on specific sound effects).

**Testing project:** [test_pr_88646.zip](https://github.com/godotengine/godot/files/14499782/test_pr_88646.zip)[^1]

[^1]: Music in the testing project by [Marc A. Pullen](https://www.youtube.com/@TheMarcAPullen).

DeeJayLSP commented:
*Something I was writing before reviews*

I added an optimized version of the reference player's logic inside AudioStreamPlaybackQOA. This makes the diff smaller and reduces a lot of unnecessary callbacks.

The patches in qoa.h are now solely meant to remove unused functions and fix errors/warnings when including it.

DeeJayLSP commented:
> There's also the alternative of decoding Ogg Vorbis/MP3 to uncompressed audio at load-time

As far as I understood, one of the motivations behind QOA's creation was to avoid having to do this.

Considering QOA is already cheap to decode, I believe it's better than sacrificing even a few KBs of memory.

DeeJayLSP commented:
Tested a web release template build (platform=web target=template_release production=yes, enabling and disabling qoa).

The size penalty of the resulting zip was only 4.7KiB.

miv391 commented on modules/qoa/audio_stream_qoa.cpp:
```diff
@@ -0,0 +1,281 @@
+		p_buffer[p_frames - todo] /= 32767.0f;
+
+		// Discard channels other than L and R
+		if (qoa_stream->channels > 2) {
+			src_index += qoa_stream->channels - 2;
+		}
```
I'm assuming that `qoa_stream->channels` stays the same during the stream's lifetime. So how about precalculating a new member variable `skipped_channels` (maybe in `instantiate_playback()`). Something like this:

```c++
skipped_channels = MAX(0, channels - 2);
```

Now the `if` could be removed here altogether:

```c++
src_index += qoa_stream.skipped_channels;
```

I'm assuming that an addition would be faster than a conditional, but as always when optimizing, some actual testing should be done.

DeeJayLSP replied:
Done.

DeeJayLSP replied:
Upon further inspection, I realized that since `src_index` is re-generated every iteration, discarding channels is not necessary at all. It's leftover of a previous buggy implementation that did require it.

DeeJayLSP committed:
Add support for QOA (Quite OK Audio) format

DeeJayLSP commented:
Ping-Pong and Backward loops successfully implemented.

fire reviewed (approved):
The last remaining criticism I had of the pull regarding missing Ping-Pong and Backward loops was solved, so I don't have any arguments against merging QOA. Maybe @reduz or the audio team can look, but we don't have an active audio team.

Approving the enhancement, didn't review the technical implementation deeply.

DeeJayLSP committed:
Loop, increment, channels

- Remove loop count, it's not present on WAV;
- Rename sign to increment;
- Remove channels in favor of stereo to match WAV;

DeeJayLSP commented:
I have changed the stream to use a `stereo` bool instead of number of channels and removed loop count. Just to increase parity with AudioStreamWAV.

Feature-wise, the only thing missing is `format`, which isn't applicable to QOA. And I don't think anyone checks it.

DeeJayLSP commented:
Superseded by #91014

akien-mga commented:
Superseded by #91014.
